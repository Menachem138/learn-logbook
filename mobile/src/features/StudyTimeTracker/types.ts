export type TimerState = 'STOPPED' | 'STUDYING' | 'BREAK';

export interface TimerDisplayProps {
  time: number;
  timerState: TimerState;
}

export interface TimerControlsProps {
  timerState: TimerState;
  onStartStudy: () => void;
  onStartBreak: () => void;
  onStop: () => void;
}

export interface TimerStatsProps {
  totalStudyTime: number;
  totalBreakTime: number;
  currentTime: number;
  timerState: TimerState;
}
