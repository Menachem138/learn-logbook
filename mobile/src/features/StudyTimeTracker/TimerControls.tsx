import React from 'react';
import { View, TouchableOpacity, Text, StyleSheet } from 'react-native';
import Icon from 'react-native-vector-icons/Feather';
import { TimerControlsProps } from './types';

export const TimerControls: React.FC<TimerControlsProps> = ({
  timerState,
  onStartStudy,
  onStartBreak,
  onStop,
}) => {
  return (
    <View style={styles.container}>
      <TouchableOpacity
        style={[
          styles.button,
          timerState === 'STUDYING' ? styles.activeButton : styles.outlineButton,
          styles.studyButton
        ]}
        onPress={onStartStudy}
        disabled={timerState === 'STUDYING'}
      >
        <Icon
          name={timerState === 'STUDYING' ? 'pause' : 'play'}
          size={16}
          color={timerState === 'STUDYING' ? '#fff' : '#16a34a'}
        />
        <Text style={[
          styles.buttonText,
          timerState === 'STUDYING' ? styles.activeButtonText : styles.outlineButtonText,
          styles.studyButtonText
        ]}>
          למידה
        </Text>
      </TouchableOpacity>

      <TouchableOpacity
        style={[
          styles.button,
          timerState === 'BREAK' ? styles.activeButton : styles.outlineButton,
          styles.breakButton
        ]}
        onPress={onStartBreak}
        disabled={timerState === 'BREAK'}
      >
        <Icon
          name={timerState === 'BREAK' ? 'pause' : 'play'}
          size={16}
          color={timerState === 'BREAK' ? '#fff' : '#ca8a04'}
        />
        <Text style={[
          styles.buttonText,
          timerState === 'BREAK' ? styles.activeButtonText : styles.outlineButtonText,
          styles.breakButtonText
        ]}>
          הפסקה
        </Text>
      </TouchableOpacity>

      <TouchableOpacity
        style={[
          styles.button,
          styles.stopButton,
          timerState === 'STOPPED' && styles.disabledButton
        ]}
        onPress={onStop}
        disabled={timerState === 'STOPPED'}
      >
        <Icon name="stop-circle" size={16} color="#fff" />
        <Text style={[styles.buttonText, styles.stopButtonText]}>
          עצור
        </Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'center',
    gap: 8,
  },
  button: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 8,
    paddingHorizontal: 16,
    borderRadius: 6,
    gap: 8,
  },
  buttonText: {
    fontSize: 14,
    fontWeight: '500',
  },
  activeButton: {
    backgroundColor: '#16a34a',
  },
  activeButtonText: {
    color: '#fff',
  },
  outlineButton: {
    backgroundColor: 'transparent',
    borderWidth: 1,
    borderColor: '#e5e7eb',
  },
  outlineButtonText: {
    color: '#374151',
  },
  studyButton: {
    borderColor: '#16a34a',
  },
  studyButtonText: {
    color: '#16a34a',
  },
  breakButton: {
    borderColor: '#ca8a04',
  },
  breakButtonText: {
    color: '#ca8a04',
  },
  stopButton: {
    backgroundColor: '#ef4444',
  },
  stopButtonText: {
    color: '#fff',
  },
  disabledButton: {
    opacity: 0.5,
  },
});
