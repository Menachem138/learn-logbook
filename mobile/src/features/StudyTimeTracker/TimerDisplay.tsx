import React from 'react';
import { View, Text, StyleSheet } from 'react-native';
import { TimerDisplayProps } from './types';

const formatTime = (ms: number): string => {
  const totalSeconds = Math.floor(ms / 1000);
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  const milliseconds = Math.floor((ms % 1000) / 10);

  return `${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}.${milliseconds.toString().padStart(2, '0')}`;
};

export const TimerDisplay: React.FC<TimerDisplayProps> = ({ time, timerState }) => {
  const getBackgroundColor = () => {
    switch (timerState) {
      case 'STUDYING':
        return '#dcfce7'; // bg-green-100
      case 'BREAK':
        return '#fef9c3'; // bg-yellow-100
      default:
        return '#f3f4f6'; // bg-gray-100
    }
  };

  return (
    <View style={styles.container}>
      <View style={[styles.timerContainer, { backgroundColor: getBackgroundColor() }]}>
        <Text style={styles.timerText}>
          {formatTime(time)}
        </Text>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
    position: 'relative',
  },
  timerContainer: {
    width: '100%',
    padding: 32,
    borderRadius: 16,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.2,
    shadowRadius: 1.41,
    elevation: 2,
  },
  timerText: {
    fontSize: 48,
    fontWeight: 'bold',
    textAlign: 'center',
    fontVariant: ['tabular-nums'],
  },
});
