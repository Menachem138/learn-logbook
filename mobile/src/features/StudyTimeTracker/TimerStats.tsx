import React from 'react';
import { View, Text, StyleSheet, Pressable } from 'react-native';
import Icon from 'react-native-vector-icons/Feather';
import { TimerStatsProps } from './types';

const formatTotalTime = (ms: number): string => {
  const totalSeconds = Math.floor(ms / 1000);
  const hours = Math.floor(totalSeconds / 3600);
  const minutes = Math.floor((totalSeconds % 3600) / 60);
  const seconds = totalSeconds % 60;

  if (hours > 0) {
    return `${hours}:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
  }
  return `${minutes}:${seconds.toString().padStart(2, '0')}`;
};

export const TimerStats: React.FC<TimerStatsProps> = ({
  totalStudyTime,
  totalBreakTime,
  currentTime,
  timerState,
}) => {
  const [activeTab, setActiveTab] = React.useState<'summary' | 'details'>('summary');
  
  const totalTime = totalStudyTime + totalBreakTime + (timerState !== 'STOPPED' ? currentTime : 0);
  const studyPercentage = totalTime > 0 
    ? ((totalStudyTime + (timerState === 'STUDYING' ? currentTime : 0)) / totalTime) * 100 
    : 0;

  return (
    <View style={styles.container}>
      <View style={styles.tabList}>
        <Pressable
          style={[styles.tab, activeTab === 'summary' && styles.activeTab]}
          onPress={() => setActiveTab('summary')}
        >
          <Text style={[styles.tabText, activeTab === 'summary' && styles.activeTabText]}>
            סיכום
          </Text>
        </Pressable>
        <Pressable
          style={[styles.tab, activeTab === 'details' && styles.activeTab]}
          onPress={() => setActiveTab('details')}
        >
          <Text style={[styles.tabText, activeTab === 'details' && styles.activeTabText]}>
            פירוט
          </Text>
        </Pressable>
      </View>

      {activeTab === 'summary' && (
        <View style={styles.summaryContent}>
          <View style={styles.summaryHeader}>
            <Text style={styles.summaryLabel}>זמן למידה</Text>
            <Text style={styles.summaryTime}>
              {formatTotalTime(totalStudyTime + (timerState === 'STUDYING' ? currentTime : 0))}
            </Text>
          </View>
          <View style={styles.progressBarContainer}>
            <View style={[styles.progressBar, { width: `${studyPercentage}%` }]} />
          </View>
          <View style={styles.progressLabels}>
            <Text style={styles.progressLabel}>0%</Text>
            <Text style={styles.progressLabel}>50%</Text>
            <Text style={styles.progressLabel}>100%</Text>
          </View>
        </View>
      )}

      {activeTab === 'details' && (
        <View style={styles.detailsContent}>
          <View style={styles.statCard}>
            <View style={[styles.statIcon, styles.studyIcon]}>
              <Icon name="book-open" size={20} color="#15803d" />
            </View>
            <View style={styles.statInfo}>
              <Text style={styles.statLabel}>זמן למידה</Text>
              <Text style={[styles.statValue, styles.studyValue]}>
                {formatTotalTime(totalStudyTime + (timerState === 'STUDYING' ? currentTime : 0))}
              </Text>
            </View>
          </View>
          <View style={styles.statCard}>
            <View style={[styles.statIcon, styles.breakIcon]}>
              <Icon name="coffee" size={20} color="#854d0e" />
            </View>
            <View style={styles.statInfo}>
              <Text style={styles.statLabel}>זמן הפסקה</Text>
              <Text style={[styles.statValue, styles.breakValue]}>
                {formatTotalTime(totalBreakTime + (timerState === 'BREAK' ? currentTime : 0))}
              </Text>
            </View>
          </View>
        </View>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
  },
  tabList: {
    flexDirection: 'row',
    marginBottom: 16,
    borderRadius: 8,
    backgroundColor: '#f3f4f6',
    padding: 2,
  },
  tab: {
    flex: 1,
    paddingVertical: 8,
    alignItems: 'center',
    borderRadius: 6,
  },
  activeTab: {
    backgroundColor: '#fff',
  },
  tabText: {
    fontSize: 14,
    color: '#6b7280',
  },
  activeTabText: {
    color: '#111827',
    fontWeight: '500',
  },
  summaryContent: {
    gap: 8,
  },
  summaryHeader: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  summaryLabel: {
    fontSize: 14,
    color: '#374151',
  },
  summaryTime: {
    fontSize: 14,
    fontWeight: '600',
    color: '#111827',
  },
  progressBarContainer: {
    height: 8,
    backgroundColor: '#f3f4f6',
    borderRadius: 4,
    overflow: 'hidden',
  },
  progressBar: {
    height: '100%',
    backgroundColor: '#22c55e',
  },
  progressLabels: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginTop: 4,
  },
  progressLabel: {
    fontSize: 12,
    color: '#6b7280',
  },
  detailsContent: {
    gap: 12,
  },
  statCard: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 12,
    borderRadius: 8,
    gap: 8,
  },
  statIcon: {
    width: 40,
    height: 40,
    borderRadius: 20,
    alignItems: 'center',
    justifyContent: 'center',
  },
  studyIcon: {
    backgroundColor: '#dcfce7',
  },
  breakIcon: {
    backgroundColor: '#fef9c3',
  },
  statInfo: {
    flex: 1,
  },
  statLabel: {
    fontSize: 14,
    color: '#374151',
  },
  statValue: {
    fontSize: 16,
    fontWeight: '600',
  },
  studyValue: {
    color: '#15803d',
  },
  breakValue: {
    color: '#854d0e',
  },
});
