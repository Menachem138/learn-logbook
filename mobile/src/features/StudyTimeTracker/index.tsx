import React from 'react';
import { View, Text, StyleSheet } from 'react-native';

const StudyTimeTracker = () => {
  return (
    <View testID="study-time-tracker" style={styles.container}>
      <View style={styles.card}>
        <View style={styles.header}>
          <Text style={styles.title}>מעקב זמן למידה</Text>
        </View>
      </View>
    </View>
  );
};

export default StudyTimeTracker;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 16,
    backgroundColor: '#f9fafb',
  },
  card: {
    width: '100%',
    maxWidth: 480,
    alignSelf: 'center',
    backgroundColor: 'rgba(255, 255, 255, 0.8)',
    borderRadius: 12,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.1,
    shadowRadius: 3,
    elevation: 3,
    borderWidth: 1,
    borderColor: '#e5e7eb',
  },
  header: {
    paddingVertical: 16,
    paddingHorizontal: 24,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    color: '#111827',
    textAlign: 'center',
  },
  content: {
    padding: 24,
    gap: 24,
  },
});
