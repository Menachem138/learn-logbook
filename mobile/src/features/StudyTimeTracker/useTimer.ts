import { useState, useRef, useCallback, useEffect } from 'react';
import { supabaseMobile } from '../../integrations/supabase/client';
import Toast from 'react-native-toast-message';
import { TimerState } from './types';
export const useTimer = (userId?: string) => {
  const [timerState, setTimerState] = useState<TimerState>('STOPPED');
  const [time, setTime] = useState<number>(0);
  const [totalStudyTime, setTotalStudyTime] = useState<number>(0);
  const [totalBreakTime, setTotalBreakTime] = useState<number>(0);
  
  const intervalRef = useRef<NodeJS.Timeout | null>(null);
  const startTimeRef = useRef<number>(0);
  const currentSessionRef = useRef<string | null>(null);
  const mountedRef = useRef(true);

  // Effect to handle cleanup on unmount
  useEffect(() => {
    return () => {
      mountedRef.current = false;
      if (intervalRef.current) {
        clearInterval(intervalRef.current);
        intervalRef.current = null;
      }
    };
  }, []);

  // Effect to handle timer updates
  useEffect(() => {
    // Skip if timer is stopped or no start time
    if (timerState === 'STOPPED' || !startTimeRef.current) {
      console.log('Timer not running:', {
        timerState,
        hasStartTime: !!startTimeRef.current,
        currentTime: Date.now()
      });
      return;
    }

    console.log('Starting timer:', {
      timerState,
      startTime: startTimeRef.current,
      currentTime: Date.now()
    });

    let frameId: number | null = null;

    // Create timer update function using requestAnimationFrame
    const updateTimer = () => {
      if (!mountedRef.current || timerState === 'STOPPED') {
        console.log('Timer update skipped:', {
          mounted: mountedRef.current,
          timerState,
          currentTime: Date.now()
        });
        return;
      }

      const now = Date.now();
      const elapsed = Math.max(0, now - startTimeRef.current);
      
      console.log('Timer tick:', {
        now,
        startTime: startTimeRef.current,
        elapsed,
        currentTime: time,
        timerState
      });

      // Update time state
      setTime(elapsed);

      // Schedule next update
      frameId = requestAnimationFrame(updateTimer);
    };

    // Start animation frame loop
    frameId = requestAnimationFrame(updateTimer);

    // Cleanup
    return () => {
      console.log('Cleaning up timer:', {
        timerState,
        time,
        hasFrame: !!frameId
      });
      
      if (frameId !== null) {
        cancelAnimationFrame(frameId);
      }
    };
  }, [timerState]);

  const showToast = useCallback((title: string, type: 'success' | 'error') => {
    Toast.show({
      type,
      text1: title,
      position: 'bottom',
      visibilityTime: 3000,
    });
  }, []);

  const clearTimerInterval = useCallback(() => {
    if (intervalRef.current) {
      clearInterval(intervalRef.current);
      intervalRef.current = null;
    }
  }, []);

  const stopTimer = useCallback(async (): Promise<boolean> => {
    if (!userId || !currentSessionRef.current) {
      console.log('Cannot stop timer: no user ID or current session');
      return false;
    }

    try {
      // Clear interval and calculate final elapsed time
      if (intervalRef.current) {
        clearInterval(intervalRef.current);
        intervalRef.current = null;
      }
      
      const elapsedTime = startTimeRef.current > 0 ? Date.now() - startTimeRef.current : 0;
      console.log('Stopping timer:', {
        elapsedTime,
        startTime: startTimeRef.current,
        currentTime: Date.now(),
        hasInterval: !!intervalRef.current
      });

      const { data, error } = await supabaseMobile
        .from('timer_sessions')
        .update({ 
          ended_at: new Date().toISOString(),
          duration: elapsedTime
        })
        .eq('id', currentSessionRef.current);

      if (error) {
        console.error('Error updating timer session:', error);
        showToast('לא ניתן לשמור את זמני הלמידה כרגע', 'error');
        return false;
      }

      if (!data) {
        console.error('No data returned from update operation');
        showToast('לא ניתן לשמור את זמני הלמידה כרגע', 'error');
        return false;
      }

      if (timerState === 'STUDYING') {
        setTotalStudyTime(prev => prev + elapsedTime);
      } else if (timerState === 'BREAK') {
        setTotalBreakTime(prev => prev + elapsedTime);
      }

      setTimerState('STOPPED');
      setTime(0);
      currentSessionRef.current = null;
      return true;
    } catch (error) {
      console.error('Error stopping timer session:', error);
      showToast('לא ניתן לשמור את זמני הלמידה כרגע', 'error');
      return false;
    }
  }, [userId, timerState, showToast, clearTimerInterval]);

  const startTimer = useCallback(async (newState: 'STUDYING' | 'BREAK'): Promise<boolean> => {
    if (!userId) {
      showToast('יש להתחבר כדי לעקוב אחר זמני הלמידה', 'error');
      return false;
    }

    try {
      // Stop current timer if running
      if (timerState !== 'STOPPED') {
        const stopped = await stopTimer();
        if (!stopped) return false;
      }

      // Create new session
      const { data: newSession, error } = await supabaseMobile
        .from('timer_sessions')
        .insert({
          user_id: userId,
          type: newState.toLowerCase(),
          started_at: new Date().toISOString(),
        })
        .select()
        .single();

      if (error) {
        console.error('Error creating session:', error);
        showToast('אירעה שגיאה בהפעלת הטיימר', 'error');
        setTimerState('STOPPED');
        return false;
      }

      if (!newSession) {
        console.error('No session data returned');
        showToast('אירעה שגיאה בהפעלת הטיימר', 'error');
        setTimerState('STOPPED');
        return false;
      }

      // Set initial state and refs
      const now = Date.now();
      startTimeRef.current = now;
      currentSessionRef.current = newSession.id;
      
      console.log('Setting initial state:', {
        newState,
        now,
        startTime: startTimeRef.current,
        currentSession: currentSessionRef.current
      });
      
      // Clear any existing interval
      if (intervalRef.current) {
        clearInterval(intervalRef.current);
        intervalRef.current = null;
      }

      // Update state
      setTimerState(newState);
      setTime(0); // Set initial time to 0
      
      // Force an immediate time update
      const initialElapsed = Math.max(0, Date.now() - startTimeRef.current);
      setTime(initialElapsed);
      
      console.log('Timer started successfully:', {
        startTime: startTimeRef.current,
        timerState: newState,
        currentTime: Date.now(),
        initialElapsed
      });
      
      return true;
    } catch (error) {
      console.error('Error starting timer:', error);
      showToast('אירעה שגיאה בהפעלת הטיימר', 'error');
      setTimerState('STOPPED');
      clearTimerInterval();
      return false;
    }
  }, [userId, timerState, stopTimer, showToast, clearTimerInterval]);

  // Cleanup on unmount
  useEffect(() => {
    return () => {
      clearTimerInterval();
    };
  }, [clearTimerInterval]);

  // Create updateTimer function that can be exposed for testing
  const updateTimer = useCallback(() => {
    if (!mountedRef.current || timerState === 'STOPPED' || !startTimeRef.current) {
      console.log('Update skipped:', {
        mounted: mountedRef.current,
        timerState,
        hasStartTime: !!startTimeRef.current
      });
      return;
    }

    const now = Date.now();
    const elapsed = Math.max(0, now - startTimeRef.current);

    console.log('Manual timer update:', {
      now,
      startTime: startTimeRef.current,
      elapsed,
      currentTime: time,
      willUpdate: elapsed !== time
    });

    if (elapsed !== time) {
      setTime(elapsed);
    }
  }, [time, timerState]);

  return {
    time,
    timerState,
    totalStudyTime,
    totalBreakTime,
    startTimer,
    stopTimer,
    updateTimer // Expose for testing
  };
};
