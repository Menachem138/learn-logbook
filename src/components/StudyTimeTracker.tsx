'use client'

import React, { useState, useEffect, useRef } from 'react';
import { Button } from '@/components/ui/button';
import { Card, CardContent, CardHeader, CardTitle } from '@/components/ui/card';
import { Tabs, TabsContent, TabsList, TabsTrigger } from '@/components/ui/tabs';
import { Progress } from '@/components/ui/progress';
import { BookOpen, Coffee, Pause, Play, StopCircle } from 'lucide-react';
import { formatTime, formatTotalTime } from '@/utils/timeUtils';
import { TimerState } from '@/types/timer';
import { useAuth } from '@/components/auth/AuthProvider';
import { supabase } from '@/integrations/supabase/client';
import { toast } from 'sonner';

export const StudyTimeTracker: React.FC = () => {
  const [timerState, setTimerState] = useState<TimerState>(TimerState.STOPPED);
  const [time, setTime] = useState<number>(0);
  const [totalStudyTime, setTotalStudyTime] = useState<number>(0);
  const [totalBreakTime, setTotalBreakTime] = useState<number>(0);
  const intervalRef = useRef<NodeJS.Timeout | null>(null);
  const { session } = useAuth();

  const startTimeRef = useRef<number>(0);
  const studyTimeRef = useRef<number>(0);
  const breakTimeRef = useRef<number>(0);
  const activeSessionRef = useRef<string | null>(null);

  useEffect(() => {
    fetchTodayStats();
    return () => {
      if (intervalRef.current) clearInterval(intervalRef.current);
    };
  }, []);

  const fetchTodayStats = async () => {
    if (!session?.user.id) return;

    const today = new Date();
    today.setHours(0, 0, 0, 0);

    try {
      const { data: sessions, error } = await supabase
        .from('timer_sessions')
        .select('*')
        .eq('user_id', session.user.id)
        .gte('created_at', today.toISOString());

      if (error) throw error;

      let studyTime = 0;
      let breakTime = 0;

      sessions?.forEach((session) => {
        if (session.type === 'study') {
          studyTime += session.duration;
        } else {
          breakTime += session.duration;
        }
      });

      setTotalStudyTime(studyTime * 1000);
      setTotalBreakTime(breakTime * 1000);
    } catch (error) {
      console.error('Error fetching timer sessions:', error);
      toast.error("אירעה שגיאה בטעינת נתוני הטיימר");
    }
  };

  const startTimer = async (state: TimerState) => {
    if (!session?.user.id) {
      toast.error("יש להתחבר כדי להשתמש בטיימר");
      return;
    }

    try {
      const { data, error } = await supabase
        .from('timer_sessions')
        .insert([
          {
            user_id: session.user.id,
            type: state === TimerState.STUDYING ? 'study' : 'break',
            duration: 0,
            started_at: new Date().toISOString(),
          },
        ])
        .select()
        .single();

      if (error) throw error;

      activeSessionRef.current = data.id;
      
      if (intervalRef.current) clearInterval(intervalRef.current);
      if (timerState !== TimerState.STOPPED) {
        const elapsedTime = Date.now() - startTimeRef.current;
        if (timerState === TimerState.STUDYING) {
          studyTimeRef.current += elapsedTime;
        } else if (timerState === TimerState.BREAK) {
          breakTimeRef.current += elapsedTime;
        }
      }
      
      setTimerState(state);
      startTimeRef.current = Date.now();
      setTime(0);
      intervalRef.current = setInterval(() => {
        setTime(prevTime => prevTime + 10);
      }, 10);

      toast.success(`התחלת ${state === TimerState.STUDYING ? 'למידה' : 'הפסקה'}`);
    } catch (error) {
      console.error('Error starting timer:', error);
      toast.error("אירעה שגיאה בהפעלת הטיימר");
    }
  };

  const stopTimer = async () => {
    if (!session?.user.id || !activeSessionRef.current) return;

    try {
      if (intervalRef.current) clearInterval(intervalRef.current);
      const elapsedTime = Date.now() - startTimeRef.current;
      
      if (timerState === TimerState.STUDYING) {
        studyTimeRef.current += elapsedTime;
      } else if (timerState === TimerState.BREAK) {
        breakTimeRef.current += elapsedTime;
      }

      const duration = Math.floor(elapsedTime / 1000);
      
      const { error } = await supabase
        .from('timer_sessions')
        .update({
          ended_at: new Date().toISOString(),
          duration: duration,
        })
        .eq('id', activeSessionRef.current);

      if (error) throw error;

      setTotalStudyTime(prev => prev + (timerState === TimerState.STUDYING ? elapsedTime : 0));
      setTotalBreakTime(prev => prev + (timerState === TimerState.BREAK ? elapsedTime : 0));
      setTimerState(TimerState.STOPPED);
      setTime(0);
      activeSessionRef.current = null;

      toast.success(`סיימת ${timerState === TimerState.STUDYING ? 'למידה' : 'הפסקה'}`);
    } catch (error) {
      console.error('Error stopping timer:', error);
      toast.error("אירעה שגיאה בעצירת הטיימר");
    }
  };

  const totalTime = totalStudyTime + totalBreakTime + (timerState !== TimerState.STOPPED ? time : 0);
  const studyPercentage = totalTime > 0 ? 
    ((totalStudyTime + (timerState === TimerState.STUDYING ? time : 0)) / totalTime) * 100 : 0;

  return (
    <Card className="w-full max-w-md mx-auto bg-white/80 backdrop-blur-sm shadow-lg border border-gray-200">
      <CardHeader className="space-y-1">
        <CardTitle className="text-center text-3xl font-bold text-gray-800">מעקב זמן למידה</CardTitle>
      </CardHeader>
      <CardContent className="space-y-6">
        <div className="relative">
          <div className={`w-full py-8 px-4 rounded-2xl ${
            timerState === TimerState.STUDYING 
              ? 'bg-green-100' 
              : timerState === TimerState.BREAK 
                ? 'bg-yellow-100' 
                : 'bg-gray-100'
          } transition-colors duration-300 ease-in-out shadow-inner`}>
            <div className="relative z-10 text-7xl font-bold text-center tabular-nums" dir="ltr">
              {formatTime(time)}
            </div>
          </div>
        </div>
        <div className="flex justify-center space-x-2">
          <Button 
            onClick={() => startTimer(TimerState.STUDYING)} 
            disabled={timerState === TimerState.STUDYING}
            variant={timerState === TimerState.STUDYING ? "default" : "outline"}
            className="transition-all duration-300 ease-in-out hover:shadow-md"
          >
            {timerState === TimerState.STUDYING ? <Pause className="mr-2 h-4 w-4" /> : <Play className="mr-2 h-4 w-4" />}
            למידה
          </Button>
          <Button 
            onClick={() => startTimer(TimerState.BREAK)} 
            disabled={timerState === TimerState.BREAK}
            variant={timerState === TimerState.BREAK ? "default" : "outline"}
            className="transition-all duration-300 ease-in-out hover:shadow-md"
          >
            {timerState === TimerState.BREAK ? <Pause className="mr-2 h-4 w-4" /> : <Play className="mr-2 h-4 w-4" />}
            הפסקה
          </Button>
          <Button 
            onClick={stopTimer} 
            disabled={timerState === TimerState.STOPPED} 
            variant="destructive"
            className="transition-all duration-300 ease-in-out hover:shadow-md"
          >
            <StopCircle className="mr-2 h-4 w-4" />
            עצור
          </Button>
        </div>
        <Tabs defaultValue="summary" className="w-full">
          <TabsList className="grid w-full grid-cols-2 mb-4">
            <TabsTrigger value="summary" className="text-sm">סיכום</TabsTrigger>
            <TabsTrigger value="details" className="text-sm">פירוט</TabsTrigger>
          </TabsList>
          <TabsContent value="summary" className="space-y-4">
            <div className="space-y-2">
              <div className="flex justify-between items-center text-sm">
                <span>זמן למידה</span>
                <span className="font-semibold">{formatTotalTime(totalStudyTime + (timerState === TimerState.STUDYING ? time : 0))}</span>
              </div>
              <Progress value={studyPercentage} className="w-full h-2" />
              <div className="flex justify-between items-center text-xs text-gray-500">
                <span>0%</span>
                <span>50%</span>
                <span>100%</span>
              </div>
            </div>
          </TabsContent>
          <TabsContent value="details" className="space-y-4">
            <div className="space-y-3">
              <div className="flex justify-between items-center bg-green-100 p-3 rounded-lg transition-all duration-300 ease-in-out hover:shadow-md">
                <span className="flex items-center text-green-800">
                  <BookOpen className="mr-2 h-5 w-5" /> זמן למידה
                </span>
                <span className="font-semibold text-green-800">
                  {formatTotalTime(totalStudyTime + (timerState === TimerState.STUDYING ? time : 0))}
                </span>
              </div>
              <div className="flex justify-between items-center bg-yellow-100 p-3 rounded-lg transition-all duration-300 ease-in-out hover:shadow-md">
                <span className="flex items-center text-yellow-800">
                  <Coffee className="mr-2 h-5 w-5" /> זמן הפסקה
                </span>
                <span className="font-semibold text-yellow-800">
                  {formatTotalTime(totalBreakTime + (timerState === TimerState.BREAK ? time : 0))}
                </span>
              </div>
            </div>
          </TabsContent>
        </Tabs>
      </CardContent>
    </Card>
  );
};