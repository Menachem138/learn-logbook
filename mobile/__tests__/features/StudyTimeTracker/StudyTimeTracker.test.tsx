import React from 'react';
import renderer from 'react-test-renderer';
import StudyTimeTracker from '../../../src/features/StudyTimeTracker/index';

jest.mock('../../../src/features/StudyTimeTracker/useTimer', () => ({
  useTimer: () => ({
    time: 0,
    timerState: 'STOPPED',
    totalStudyTime: 0,
    totalBreakTime: 0,
    startTimer: jest.fn(),
    stopTimer: jest.fn(),
  }),
}));

describe('StudyTimeTracker', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<StudyTimeTracker />).toJSON();
    expect(tree).toBeTruthy();
  });

  it('matches snapshot', () => {
    const tree = renderer.create(<StudyTimeTracker />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
