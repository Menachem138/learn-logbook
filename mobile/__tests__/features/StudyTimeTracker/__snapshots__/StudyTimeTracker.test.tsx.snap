// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudyTimeTracker matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#f9fafb",
      "flex": 1,
      "padding": 16,
    }
  }
  testID="study-time-tracker"
>
  <View
    style={
      {
        "alignSelf": "center",
        "backgroundColor": "rgba(255, 255, 255, 0.8)",
        "borderColor": "#e5e7eb",
        "borderRadius": 12,
        "borderWidth": 1,
        "elevation": 3,
        "maxWidth": 480,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 3,
        "width": "100%",
      }
    }
  >
    <View
      style={
        {
          "paddingHorizontal": 24,
          "paddingVertical": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#111827",
            "fontSize": 24,
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        מעקב זמן למידה
      </Text>
    </View>
  </View>
</View>
`;
